File: src/odin-libraries/python/odin_logger.py
from pymongo import MongoClient
from os import environ

client = MongoClient(environ.get('ODIN_MONGODB'))
# get odin db
db = client['odin']
collection = db['observability']

class OdinLogger:
    @classmethod
    def log(self, type, desc, value, id, collection=collection):
        self.find_and_insert(collection, type, desc, value,  id)
    
    @staticmethod
    def find_and_insert(collection, type, desc, value, id):
        collection.update_one(
            {'id': str(id), 'desc':str(desc), 'type': str(type)},
            {'$set' : {
                'type' : str(type),
                'desc' : str(desc),
                'value' : str(value),
                'id' : str(id),
            }}, upsert=True
        )

File: src/odin-libraries/nodejs/odin/odinLogger.js
const MongoClient = require('mongodb').MongoClient;
const assert = require('assert')

// Connection URL
const url = (process.env.ODIN_MONGODB)

class OdinLogger {
    constructor() {
        MongoClient.connect(url, function(err, client){
            assert.equal(null, err);
            if (err) {
                this.success = false;
            }
            this.db = client.db('odin');
            this.collection = this.db.collection('observability');
            this.success = true;
        })
    }

    async log(type, desc, value, id, timestamp, collection=this.collection){
        if (this.success == true){
            await OdinLogger.insert(collection, type, desc, value, id, timestamp)
        }
    }

    static async insert(collection, type, desc, value, id, timestamp){
        await collection.insertOne({
            'type' : String(type),
            'desc' : String(desc),
            'value' : String(value),
            'id' : String(id),
            'timestamp' : String(timestamp)
        })
    }
}

module.exports.OdinLogger = OdinLogger;